test/robot.go

```diff
@@ -2,215 +2,155 @@ package test
-}
-
-type Position struct {
-	X,
-	Y int
-}
-
-func (p Position) Equals(pos Position) bool {
-	return p.X == pos.X && p.Y == pos.Y
-}
-
-type TypedPosition struct {
-	Position
-	Type string
-}
 
-func (p TypedPosition) move(d int) TypedPosition {
-	tp := p
-	tp.Type = "MOVE"
-	switch d {
-	case up:
-		tp.X--
-	case down:
-		tp.X++
-	case left:
-		tp.Y--
-	case right:
-		tp.Y++
+	var body []byte
+	if body, err = json.Marshal(tr.Moves); err != nil {
+		return
 	}
-	return tp
+	return ws.Write(sub.Key.Parent(), body)
 }
 
-func (move *TypedPosition) validate(m Map) bool {
-	if move.X < 0 || move.Y < 0 || move.X > m.Max.X || move.X > m.Max.Y {
-		move.Type = "OUT_OF_BOUNDS"
-		return false
-	}
-	for _, val := range m.Obstacles {
-		if move.X == val.X && move.Y == val.Y {
-			move.Type = "HIT_OBSTACLE"
-			return false
+func testRobot(m *model.Map, in io.Reader) (tr model.RobotTestResults, err error) {
+	tr.Moves = append(tr.Moves, model.RobotLogEntry{Position: m.Start, Event: model.Move})
+	direction := model.Right
+	r := bufio.NewReader(in)
+	var picks []int
+	for {
+		current, err := r.ReadBytes('\n')
+		if len(current) == 0 {
+			break
+		} else if err != nil && err != io.EOF {
+			return tr, err
 		}
-	}
-	return true
-}
 
-type Moves []TypedPosition
+		current = bytes.ToLower(current)
+		current = bytes.TrimSuffix(current, []byte("\n"))
 
-func (m Moves) last() TypedPosition {
-	return m[len(m)-1]
-}
-
-func (mv Moves) validate(pos int, m Map) bool {
-	return (&mv[pos]).validate(m)
-}
-
-type Map struct {
-	Start     Position
-	Finish    Position
-	Coins     []Position
-	Obstacles []Position
-	Max       Position
-}
-
-const (
-	up = iota
-	right
-	down
-	left
-)
-
-func turnLeft(d int) int {
-	d--
-	if d < 0 {
-		d = left
-	}
-	return d
-}
-
-func turnRight(d int) int {
-	d++
-	if d > 3 {
-		d = up
-	}
-	return d
-}
-
-func testRobot(m Map, in string) (moves []TypedPosition, err error) {
-	moves = append(moves, TypedPosition{Position: m.Start, Type: "MOVE"})
-
-	in = strings.ToLower(in)
-	commands := strings.Split(in, "\n")
-
-	direction := right
-
-	for counter, val := range commands {
-		c := strings.Split(val, " ")
-		switch c[0] {
+		c := bytes.Split(current, []byte(" "))
+		switch string(c[0]) {
 		case "move":
-			n, err := strconv.Atoi(c[1])
+			if len(c) < 2 {
+				return tr, errors.New("move missing parameter")
+			}
+
+			n, err := strconv.Atoi(string(c[1]))
 			if err != nil {
-				return moves, errors.New("bad move argument")
+				return tr, err
 			}
-			nextIsPick := !((counter < len(commands)-1 && commands[counter+1] != "PICK") || counter > len(commands))
-			var ok bool
-			moves, ok = move(moves, n, direction, m, nextIsPick)
-			if !ok {
-				return moves, nil
+
+			if ok := move(&tr, n, direction, m); !ok {
+				return tr, nil
 			}
 		case "right":
 			direction = turnRight(direction)
 		case "left":
 			direction = turnLeft(direction)
 		case "pick":
-			moves = pick(moves, m.Coins)
+			picks = append(picks, len(tr.Moves)-1)
```

Comment:
Here you do not append anything to `tr.Moves`, so when iterating through `tr.Maps` to check the picks later on, you'll likely override some other elements `Event` property.